src/protocol.ts

```diff
@@ -27,12 +27,66 @@ export async function protocolWrite(stream: Writable, data: AbqTypes.ManifestMes
   })
 }
 
+/**
+ * Reads a single message using the standard 4-byte header protocol.
+ *
+ * Combining this with the `protocolReader` method will have unexpected consequences
+ *
+ * returns null if all messages have been read / testing is over
+ */
+export async function protocolRead(stream: Readable): Promise<AbqTypes.InitMessage | AbqTypes.TestCaseMessage | null> {
+  return await new Promise(resolve => {
+    // inspired by https://github.com/dex4er/js-promise-readable/blob/master/src/promise-readable.ts#L25
+    if (!stream.readable || stream.closed || stream.destroyed) {
+      return resolve(null)
+    }
+
+    const readableHandler = () => {
+      const messageSizeBuffer = stream.read(4)
```

Comment:
I _believe_ the socket should always have the whole message ready for consumption but I'm not positive. Will play with it

Reply:
I wonder if we need to move `messageSize` outside of `readableHandler`, then unset it once we've successfully `socket.read(messageSize)`? I'm not sure how likely this is, but I think this code would consume the size but drop the message, and next time around try to read the first `4` of the message itself and interpret it as the size.

Reply:
if that's the case, is `stream.read(4)` safe in the first place or do we _have_ to read chunks?

Reply:
According to the docs, `read(size)` will either return a message of `size` or null, so I think this can be done safely.